In [ ]:
"""
Bài 1. Viết chương trình nhập vào một chuỗi
Xuất ra màn hình chuỗi ký tự hoa
Xuất ra màn hình chuỗi ký tự có ký tự đầu là ký tự hoa
Xuất ra màn hình chuỗi ký tự thường
"""


In [ ]:
st = input("Enter your string:")
st_upper = st.upper()
st_lower = st.lower()
st_title = st.title()

print("A Upper String is: " + st_upper)
print("A Lower String is : " + st_lower)
print("A Title String is : " + st_title)

In [ ]:
'''
Bài 2. Cho chuỗi biểu diễn địa chỉ email:   " minhnhutvh@gmail.com" 
Rút trích và hiển thị chuỗi "gmail.com" (Đây chính là  tên Host)
'''

In [ ]:
data = "minhnhutvh@gmail.com"
position = data.find("@")               
host = data[position +1 :]
print(host)


In [ ]:
'''
Bài 3. Cho chuỗi biểu diễn:  
        ‘minhnhutvh@gmai.com Sat Jan 5 09:14:16’
Rút trích và hiển thị chuỗi “gmail.com”
'''

In [ ]:
data = "minhnhutvh@gmail.com Sat Jan 5 09:14:16"
Start_position = data.find("@")                #  10
End_position = data.find(" ", Start_position)  #  26
host = data[Start_position +1 : End_position]
print(host)


In [ ]:
''' Những bài thực hành còn lại. Sinh viên tự làm'''

In [ ]:
# Ex4

st1 = input("Enter your string: ")
st2 = input(" Enter the word that needs to replace: ")
st3 = input(" Enter the word after being replaced: ")

st1 = st1.replace(st2, st3)
print(st1)

In [ ]:
# Ex5

st = input("Enter your string: ")
First_five_characters = st[:5]
Last_five_characters = st[len(st) - 5 :]
Str_1_line = 4 * (st + " ")
Str_4_line = 4 * (st + "\n")

print("First five characters are: " + First_five_characters)
print("Last five characters are: " + Last_five_characters)
print("Four strings of one line are: " + Str_1_line)
print("Four strings of four line are: \n" + Str_4_line)

In [4]:
# Ex6

from math import pi, pow
from utils.formatInput import floatInput

r, h = floatInput("Enter r:"), floatInput("Enter h:")
V = pi * pow(r, 2) * h

print(f"V(r={r}, h={h}) =", V)

V(r=3.0, h=5.0) = 141.3716694115407


In [ ]:
# Ex7

from math import sin, degrees, pi
from utils.formatInput import floatInput as fI


def isValidTriangle(
    edge_a: float,
    edge_b: float,
    edge_c: float,
    angle_A: float,
    angle_B: float,
    angle_C: float,
):
    edge_greater_then_zero = edge_a > 0 and edge_b > 0 and edge_c > 0
    print("edge_greater_then_zero: ", edge_greater_then_zero)

    angle_in_valid_range = 0 < angle_A < pi and 0 < angle_B < pi and 0 < angle_C < pi
    print("angle_in_valid_range: ", angle_in_valid_range)

    # góc input đơn vị rad, dùng degrees thành float có thể bị lệch khi so trực tiếp với 180, so sánh hiệu với float nhỏ
    total_angle_is_180 = (
        abs(degrees(angle_A) + degrees(angle_B) + degrees(angle_C) - 180) < 0.0000001
    )
    print("total_angle_is_180: ", total_angle_is_180, '\n')

    return edge_greater_then_zero and angle_in_valid_range and total_angle_is_180


def triangleS(edge_1: float, edge_2: float, mid_angle: float):
    result = 1 / 2 * edge_1 * edge_2 * sin(mid_angle)
    print(
        f"triangleS(edge_1={edge_1}, edge_2={edge_2}, mid_angle={mid_angle}) =", result
    )
    return result


"""
Test
a: 2
b: 2
c: 2
A: 1.0471975512
B: 1.0471975512
C: 1.0471975512
"""
edge_a, edge_b, edge_c = fI("edge a:"), fI("edge b:"), fI("edge c:")
angle_A, angle_B, angle_C = fI("angle A (rad):"), fI("angle B (rad):"), fI("angle C (rad):")

if not isValidTriangle(edge_a, edge_b, edge_c, angle_A, angle_B, angle_C):
    raise ValueError("Invalid triangle")

triangleS(edge_a, edge_b, angle_C)
triangleS(edge_a, edge_c, angle_B)
triangleS(edge_b, edge_c, angle_A)

edge_greater_then_zero:  True
angle_in_valid_range:  True
total_angle_is_180:  True 

triangleS(edge_1=2.0, edge_2=2.0, mid_angle=1.0471975512) = 1.7320508075722796
triangleS(edge_1=2.0, edge_2=2.0, mid_angle=1.0471975512) = 1.7320508075722796
triangleS(edge_1=2.0, edge_2=2.0, mid_angle=1.0471975512) = 1.7320508075722796


1.7320508075722796

In [ ]:
# Ex8

from math import sqrt
from utils.formatInput import floatInput

def equilateralTriangleS(a: float):
    if (a < 0): raise ValueError('Require positive values')
    return a ** 2 * (sqrt(3) / 4)

edge_c = floatInput('Enter edge length:')
print(f'equilateralTriangleS({edge_c}) =', equilateralTriangleS(edge_c))


equilateralTriangleS(2.0) = 1.7320508075688772


# Ex9

- Run Ex7 with:

```
a: 2
b: 2
c: 2
A: 1.0471975512
B: 1.0471975512
C: 1.0471975512
```
result: 1.7320508075722796

- Run Ex8 with edge = 2:
result: 1.7320508075722796

-> The results are equal.

In [ ]:
# Ex10

from utils.formatInput import floatInput as fI

width, height = fI("Enter width: "), fI("Enter height: ")

if width <= 0 or height <= 0:
    raise ValueError("Invalid size value")

print(f"S_room(width={width}, height={height}) =", width * height)

S_room(width=2.0, height=5.0) = 10.0


In [ ]:
# Ex11

from utils.formatInput import floatInput as fI

SQUARE_METER_TO_ACRE = 43560

width, height = fI("Enter width (Unit: m): "), fI("Enter height (Unit: m): ")

if width <= 0 or height <= 0:
    raise ValueError("Invalid size value")

print(f"S_field(width={width}, height={height}) =", width * height / 43560, "(Acre)")

S_field(width=10000.0, height=200.0) = 45.91368227731864 (Acre)


In [ ]:
# Ex12

from utils.formatInput import floatInput

TIP_PERCENT = 18 / 100

# Không rõ yêu cầu đề, giả sử tiền thuế tính lên cả tiền tip
TAX_PERCENT = (1 + TIP_PERCENT) * 5 / 100

cost = floatInput("Enter meal cost:")
if cost <= 0:
    raise ValueError("Invalid cost value")

tax = round(cost * TAX_PERCENT, 2)
tip = round(cost * TIP_PERCENT, 2)

print("Food cost:", cost)
print("-\tTax:", tax)
print("-\tTip:", tip)
print("-\tTotal meal cost:", round(cost + tip + tax, 2))

Food cost: 10101.0
-	Tax: 595.96
-	Tip: 1818.18
-	Total meal cost: 12515.14


In [ ]:
# Ex13

from utils.formatInput import intInput

n = intInput("Enter n:")
if n <= 0:
    raise ValueError("Invalid N+ value")

print(f"Sum(1 to {n}) =", int((n * (n + 1)) / 2))

Sum(1 to 10) = 55


In [ ]:
# Ex14

from math import log10
from utils.formatInput import intInput

a, b = intInput("Enter integer a: "), intInput("Enter integer b: ")

print(f"a = {a}, b = {b}")
print("- Sum =", a + b)
print("- Difference =", a - b)
print("- Product =", a * b)

# div/mod 0 not defined
if b != 0:
    print("- Division =", a / b)
    print("- Remainder =", a % b)
else:
    print("- Division and remainder not defined (b is zero)")

# log10(a) valid when a > 0
if a > 0:
    print("- log10 =", log10(a))
else:
    print("- log10(a) not defined (a <= 0)")

print("- Power =", a**b)

a = 26, b = 8
- Sum = 34
- Difference = 18
- Product = 208
- Division = 3.25
- Remainder = 2
- log10 = 1.414973347970818
- Power = 208827064576


In [12]:
# Ex15

from utils.formatInput import floatInput

C_WATER = 4.186
COST_PER_KWH = 8.9
JOULES_TO_KWH = 2.777e-7


def calc_Q(m: float, c: float, deltaT: float):
    result = m * c * deltaT
    if result < 0:
        raise Exception("Invalid input")
    return result


def calc_cost(Q: float):
    return Q * JOULES_TO_KWH * COST_PER_KWH


m, deltaT = floatInput("Enter m (g): "), floatInput("Enter deltaT (Celsius): ")

Q = calc_Q(m, C_WATER, deltaT)
print("Q (J) =", round(Q, 3))
print("Cost (cent) =", round(calc_cost(Q), 3))

Q (J) = 3265080.0
Cost (cent) = 8.07


In [18]:
# Ex16

from utils.formatInput import floatInput
from math import sqrt

G = 9.8  # m/s^2
V_ZERO = 0  # m/s


def calc_free_fall_v(h: float, v_zero: float = V_ZERO):
    return sqrt(v_zero**2 + 2 * G * h)


print("Speed when hits the ground (m/s):", calc_free_fall_v(floatInput("Enter height (m):")))

Speed when hits the ground (m/s): 28.0


In [ ]:
# Ex17

from utils.formatInput import floatInput


def calc_WCI(temp_c: float, wind_speed_kmh: float) -> int:
    wci = (
        13.12
        + 0.6215 * temp_c
        - 11.37 * (wind_speed_kmh**0.16)
        + 0.3965 * temp_c * (wind_speed_kmh**0.16)
    )
    return round(wci)


# Nhập dữ liệu
Ta = floatInput("Enter air temperature (Celsius): ")
V = floatInput("Enter wind speed V (km/h): ")

# Tính WCI
WCI = calc_WCI(Ta, V)

print(f"Wind Chill Index (Ta={Ta}°C, V={V} km/h) = {WCI}")